### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] false positive when parameter is used in inferred mapped return type

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#9667]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&showAST=es&fileType=.tsx&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwFsoAHYkYAJRA2RlQBUBPMgHnvhAA8MRVgBnePwwwsqAOYA%2BABScAXPHoBKBQG94AbQDS8MYoC6CgMIksGKBCwAvECy2T4AXwBQzlOmx5dqRCBhxgAFkSMkpqWgZmW3YuHj5BYVEJGXlFJXhVZ3h4OBo6QhDyKjzIslklAG5nF2cAN1hvX39yeABeRr8A4NIi8LomMoAicV4-LDBdQWR%2BckGlZyA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJADuIppHBgAviF1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
declare function mappedReturnType<T extends string>(x: T): { [K in T]: Capitalize<K> }

function inferredMappedReturnType<T extends string>(x: T) {
  return mappedReturnType(x);
}

var inferred = inferredMappedReturnType("generic is used")
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-parameters": "warn"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I expected no error

### Actual Result

Error on the type parameter in `inferredMappedReturnType` - "Type parameter T is used only once".

**Comment by kirkwaiblinger** [comment]
Related: https://github.com/typescript-eslint/typescript-eslint/issues/9524
CC: @danvk
